WIP to switch to using noise pipes

ara4n opened the pull request:
~~doesn't remotely work yet.~~

ready for review

ara4n committed:
blind implementation of noise pipes

ara4n committed:
use new noise API

ara4n committed:
apply go fmt

ara4n committed:
add keystores

ara4n committed:
make it build

ara4n committed:
improve doc and errors

ara4n committed:
make noisepipes work

ara4n committed:
Merge branch 'master' into matthew/noise2

ara4n committed:
fix up debugging & commentary

ara4n committed:
more commentary

ara4n committed:
go fmt

ara4n commented:
ready for review.

ara4n committed:
s/receiver/responder/ for noise terminology

ara4n committed:
spell out handshake retry reqts

ara4n committed:
remove dummy RNG

ara4n committed:
skip nicely if decryption fails

ara4n committed:
make state machine states explicit and less fragile

ara4n committed:
handle races with handshakes when sending msgs (in theory)

erikjohnston commented on noise.go:
```diff
@@ -0,0 +1,455 @@
+	// See also https://moderncrypto.org/mail-archive/noise/2018/001921.html
+
+	if msg != nil {
+		ns.msgQueue = append(ns.msgQueue, msg)
```
You'll probably need a lock for this, as I assume `EncryptAndSendMessage` can be called from different goroutines? Also, should we return here?

ara4n committed:
oops, we use queues now

ara4n committed:
all the TODOs for explaining how to handle retries properly.

This probably obsoletes the https://github.com/matrix-org/go-coap/pull/4 branch

babolivier committed:
Minimal untested version of retries

babolivier committed:
Copy the read data back into m

babolivier committed:
Test if message is handshake within the scope of 'if srv.Encryption'

babolivier committed:
Refactorings

babolivier committed:
Improve logging (mostly)

babolivier committed:
Add type RetriesQueue

babolivier committed:
Improve handling of XX3 handshake steps

babolivier committed:
Move retries queue to RetriesQueue, start implementing buffer for handshakes

babolivier committed:
Use retries queue's seqnum instead of conn's

ara4n committed:
fix deadlock

babolivier committed:
Port Matthew's changes over to go-coap

babolivier committed:
Move buffering of msgs during HS to the retries layer

babolivier committed:
Actually retry

babolivier committed:
Use right conn when retrying

babolivier committed:
Typo

babolivier committed:
Process pending messages when handshake is over

babolivier committed:
Fix crashes on timeout

babolivier committed:
Reset connections on errors and use app-defined values for retries

babolivier committed:
Don't reset connections to nil addresses

babolivier committed:
Reset conns to nil addresses in a smarter way

babolivier committed:
Implement re-handshake dance on receiving incoherent message
